File: sdk/typescript/src/namespaces/receipts.ts
/**
 * Receipts Namespace API
 *
 * Provides a namespaced interface for decision receipt management.
 * Critical for SME compliance, audit trails, and defensible decision-making.
 */

import type {
  DecisionReceipt,
  GauntletReceiptExport,
  PaginationParams,
} from '../types';

// Re-export types from ../types for convenience
export type { DecisionReceipt, GauntletReceiptExport } from '../types';

/**
 * Interface for the internal client methods used by ReceiptsAPI.
 */
interface ReceiptsClientInterface {
  request<T = unknown>(method: string, path: string, options?: Record<string, unknown>): Promise<T>;
  listGauntletReceipts(params?: { verdict?: string } & PaginationParams): Promise<{ receipts: DecisionReceipt[] }>;
  getGauntletReceipt(receiptId: string): Promise<DecisionReceipt>;
  verifyGauntletReceipt(receiptId: string): Promise<{ valid: boolean; hash: string }>;
  exportGauntletReceipt(receiptId: string, format: 'json' | 'html' | 'markdown' | 'sarif'): Promise<GauntletReceiptExport>;
}

/**
 * Receipts API namespace.
 *
 * Provides methods for managing decision receipts:
 * - Gauntlet receipt listing and retrieval
 * - Verify receipt integrity (cryptographic hash)
 * - Export receipts in various formats
 * - Access findings and dissenting views
 *
 * Decision receipts provide audit-ready documentation of AI decisions,
 * essential for compliance, governance, and defensible decision-making.
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai', apiKey: 'your-key' });
 *
 * // List gauntlet receipts
 * const { receipts } = await client.receipts.listGauntlet();
 *
 * // Get a specific gauntlet receipt
 * const receipt = await client.receipts.getGauntlet('receipt-123');
 *
 * // Verify receipt integrity
 * const { valid, hash } = await client.receipts.verifyGauntlet('receipt-123');
 *
 * // Export as HTML for stakeholder review
 * const html = await client.receipts.exportGauntlet('receipt-123', 'html');
 * ```
 */
export class ReceiptsAPI {
  constructor(private client: ReceiptsClientInterface) {}

  // ===========================================================================
  // Generic Receipt Methods
  // ===========================================================================

  /**
   * List receipts with optional filtering.
   */
  async list(params?: { verdict?: string } & PaginationParams): Promise<{ receipts: DecisionReceipt[] }> {
    return this.client.listGauntletReceipts(params);
  }

  /**
   * Get a receipt by ID.
   */
  async get(receiptId: string): Promise<DecisionReceipt> {
    return this.client.getGauntletReceipt(receiptId);
  }

  /**
   * Verify a receipt's integrity.
   */
  async verify(receiptId: string): Promise<{ valid: boolean; hash: string }> {
    return this.client.verifyGauntletReceipt(receiptId);
  }

  /**
   * Verify a receipt with full signature validation.
   */
  async verifyFull(receiptId: string): Promise<Record<string, unknown>> {
    return this.client.request('GET', `/api/gauntlet/receipts/${encodeURIComponent(receiptId)}/verify-full`);
  }

  // ===========================================================================
  // Gauntlet Receipts
  // ===========================================================================

  /**
   * List gauntlet receipts with optional filtering.
   *
   * Gauntlet receipts are generated from attack/defend stress tests.
   */
  async listGauntlet(params?: { verdict?: string } & PaginationParams): Promise<{ receipts: DecisionReceipt[] }> {
    return this.client.listGauntletReceipts(params);
  }

  /**
   * Get a gauntlet receipt by ID.
   */
  async getGauntlet(receiptId: string): Promise<DecisionReceipt> {
    return this.client.getGauntletReceipt(receiptId);
  }

  /**
   * Verify a gauntlet receipt's integrity.
   */
  async verifyGauntlet(receiptId: string): Promise<{ valid: boolean; hash: string }> {
    return this.client.verifyGauntletReceipt(receiptId);
  }

  /**
   * Export a gauntlet receipt in various formats.
   *
   * @param receiptId - The receipt ID to export
   * @param format - Export format:
   *   - json: Machine-readable JSON
   *   - markdown: Human-readable Markdown
   *   - html: Styled HTML document
   *   - sarif: SARIF format for security tooling
   *
   * @example
   * ```typescript
   * // Export as HTML for stakeholder review
   * const html = await client.receipts.exportGauntlet('receipt-123', 'html');
   *
   * // Export as SARIF for security integration
   * const sarif = await client.receipts.exportGauntlet('receipt-123', 'sarif');
   * ```
   */
  async exportGauntlet(
    receiptId: string,
    format: 'json' | 'markdown' | 'html' | 'sarif'
  ): Promise<GauntletReceiptExport> {
    return this.client.exportGauntletReceipt(receiptId, format);
  }

  // ===========================================================================
  // Helpers
  // ===========================================================================

  /**
   * Check if a receipt has any dissenting views.
   */
  hasDissent(receipt: DecisionReceipt): boolean {
    return (receipt.dissenting_agents?.length ?? 0) > 0;
  }

  /**
   * Get the consensus status from a receipt.
   */
  getConsensusStatus(receipt: DecisionReceipt): {
    reached: boolean;
    confidence: number;
    participatingAgents: number;
    dissentingAgents: number;
  } {
    return {
      reached: receipt.consensus_reached ?? false,
      confidence: receipt.confidence ?? 0,
      participatingAgents: receipt.participating_agents?.length ?? 0,
      dissentingAgents: receipt.dissenting_agents?.length ?? 0,
    };
  }
}

File: sdk/python/aragora_sdk/namespaces/receipts.py
"""
Receipts Namespace API

Provides methods for decision receipt management:
- List and retrieve receipts
- Verify receipt integrity
- Export in various formats
- Share receipts
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient

_List = list  # Preserve builtin list for type annotations

ExportFormat = Literal["json", "html", "markdown", "pdf", "sarif", "csv"]


class ReceiptsAPI:
    """
    Synchronous Receipts API.

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> receipts = client.receipts.list_gauntlet(verdict="PASS")
        >>> for receipt in receipts["results"]:
        ...     print(receipt["receipt_id"], receipt["confidence"])
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    # =========================================================================
    # Gauntlet Receipts / Results
    # =========================================================================

    def list_gauntlet(
        self,
        verdict: str | None = None,
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List recent gauntlet results (from attack/defend stress tests).

        Args:
            verdict: Filter by verdict
            limit: Maximum results
            offset: Pagination offset

        Returns:
            List of gauntlet results
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if verdict:
            params["verdict"] = verdict
        return self._client.request("GET", "/api/v2/gauntlet/results", params=params)

    def get_gauntlet(self, receipt_id: str) -> dict[str, Any]:
        """
        Get a gauntlet receipt by gauntlet ID.

        Args:
            receipt_id: Receipt ID

        Returns:
            Gauntlet receipt details
        """
        return self._client.request("GET", f"/api/v2/gauntlet/{receipt_id}/receipt")

    def verify_gauntlet(self, receipt_id: str) -> dict[str, Any]:
        """
        Verify a gauntlet receipt's integrity.

        Args:
            receipt_id: Receipt ID

        Returns:
            Verification result
        """
        return self._client.request(
            "POST",
            f"/api/v2/gauntlet/{receipt_id}/receipt/verify",
        )

    def export_gauntlet(
        self,
        receipt_id: str,
        format: Literal["json", "html", "markdown", "sarif"] = "json",
    ) -> dict[str, Any]:
        """
        Export a gauntlet receipt.

        Args:
            receipt_id: Receipt ID
            format: Export format (json, html, markdown, sarif)

        Returns:
            Exported receipt data
        """
        format_value = "md" if format == "markdown" else format
        return self._client.request(
            "GET",
            f"/api/v2/gauntlet/{receipt_id}/receipt",
            params={"format": format_value},
        )

    # =========================================================================
    # Helper Methods
    # =========================================================================

    @staticmethod
    def has_dissent(receipt: dict[str, Any]) -> bool:
        """
        Check if a receipt has any dissenting views.

        Args:
            receipt: Receipt data

        Returns:
            True if there are dissenting agents
        """
        dissenting = receipt.get("dissenting_agents", [])
        return len(dissenting) > 0

    @staticmethod
    def get_consensus_status(receipt: dict[str, Any]) -> dict[str, Any]:
        """
        Get the consensus status from a receipt.

        Args:
            receipt: Receipt data

        Returns:
            Consensus status with reached, confidence, and agent counts
        """
        return {
            "reached": receipt.get("consensus_reached", False),
            "confidence": receipt.get("confidence", 0.0),
            "participating_agents": len(receipt.get("participating_agents", [])),
            "dissenting_agents": len(receipt.get("dissenting_agents", [])),
        }


class AsyncReceiptsAPI:
    """
    Asynchronous Receipts API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     results = await client.receipts.list_gauntlet(verdict="PASS")
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    # =========================================================================
    # Gauntlet Receipts / Results
    # =========================================================================

    async def list_gauntlet(
        self,
        verdict: str | None = None,
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List recent gauntlet results (from attack/defend stress tests)."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if verdict:
            params["verdict"] = verdict
        return await self._client.request("GET", "/api/v2/gauntlet/results", params=params)

    async def get_gauntlet(self, receipt_id: str) -> dict[str, Any]:
        """Get a gauntlet receipt by gauntlet ID."""
        return await self._client.request(
            "GET",
            f"/api/v2/gauntlet/{receipt_id}/receipt",
        )

    async def verify_gauntlet(self, receipt_id: str) -> dict[str, Any]:
        """Verify a gauntlet receipt's integrity."""
        return await self._client.request(
            "POST",
            f"/api/v2/gauntlet/{receipt_id}/receipt/verify",
        )

    async def export_gauntlet(
        self,
        receipt_id: str,
        format: Literal["json", "html", "markdown", "sarif"] = "json",
    ) -> dict[str, Any]:
        """Export a gauntlet receipt."""
        format_value = "md" if format == "markdown" else format
        return await self._client.request(
            "GET",
            f"/api/v2/gauntlet/{receipt_id}/receipt",
            params={"format": format_value},
        )
